modules/siddhi-extensions/markov-models/src/main/java/org/wso2/siddhi/extension/markov/MarkovChainStreamProcessor.java

```diff
@@ -0,0 +1,280 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.siddhi.extension.markov;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.wso2.siddhi.core.config.ExecutionPlanContext;
+import org.wso2.siddhi.core.event.ComplexEvent;
+import org.wso2.siddhi.core.event.ComplexEventChunk;
+import org.wso2.siddhi.core.event.stream.StreamEvent;
+import org.wso2.siddhi.core.event.stream.StreamEventCloner;
+import org.wso2.siddhi.core.event.stream.populater.ComplexEventPopulater;
+import org.wso2.siddhi.core.executor.ConstantExpressionExecutor;
+import org.wso2.siddhi.core.executor.ExpressionExecutor;
+import org.wso2.siddhi.core.query.processor.Processor;
+import org.wso2.siddhi.core.query.processor.SchedulingProcessor;
+import org.wso2.siddhi.core.query.processor.stream.StreamProcessor;
+import org.wso2.siddhi.core.util.Scheduler;
+import org.wso2.siddhi.query.api.definition.AbstractDefinition;
+import org.wso2.siddhi.query.api.definition.Attribute;
+import org.wso2.siddhi.query.api.exception.ExecutionPlanValidationException;
+
+/**
+ * markovChain( id, state, durationToKeep, alertThreshold, notificationsHoldLimit/markovMatrixStorageLocation, train )
+ * Returns last state, transition probability and notification
+ * Accept Type(s): STRING, STRING, INT/LONG/TIME, DOUBLE, (INT/LONG, STRING), BOOLEAN
+ * Return Type: STRING, DOUBLE, BOOLEAN
+ */
+public class MarkovChainStreamProcessor extends StreamProcessor implements SchedulingProcessor {
+
+    private enum TrainingMode {
+        PREDEFINED_MATRIX, REAL_TIME
+    }
+
+    private Scheduler scheduler;
+    private TrainingMode trainingMode;
+    private long durationToKeep;
+    private double alertThresholdProbability;
+    private long notificationsHoldLimit;
+    private String markovMatrixStorageLocation;
+    private Boolean trainingOption;
+    private ExpressionExecutor trainingOptionExpressionExecutor;
+    private MarkovChainTransitionProbabilitiesCalculator markovChainTransitionProbabilitiesCalculator;
+    private long lastScheduledTime;
+
+    /**
+     * The init method of the MarkovChainStreamProcessor,
+     * this method will be called before other methods
+     *
+     * @param inputDefinition the incoming stream definition
+     * @param attributeExpressionExecutors the executors of each function parameters
+     * @param executionPlanContext the context of the execution plan
+     * @return the additional output attributes introduced by the function
+     */
+    @Override
+    protected List<Attribute> init(AbstractDefinition inputDefinition,
```

Comment:
This method looks too long. Better to restructure by moving pieces to private methods. e.g. durationObject.